xds client implementation

jyotimahapatra opened the pull request:
Signed-off-by: Jyoti Mahapatra <jmahapatra@lyft.com>

jyotimahapatra commented on internal/app/upstream/client.go:
```diff
@@ -55,11 +74,177 @@ type Response struct {
+}
+
+// NewMockClient creates a mock implementation for testing
+func NewMockClient(
```
What are the best practices for exposing Mock functions?

lita replied:
There are multiple ways to do it. What you are doing is fine. The other way I have seen is moving this method to a fake.go file that has access to all the private fields/methods.

lita commented on internal/app/upstream/client.go:
```diff
@@ -55,11 +74,177 @@ type Response struct {
+	}
+
+	signal := make(chan *version, 1)
+	signal <- &version{}
```
Does xDS protocol require an empty version to start? Not clear why signal needs to buffered but response does not. It would be good to add a comment here.

jyotimahapatra replied:
I intended both to be [unbufferred](https://golang.org/doc/effective_go.html#channels). I'll correct it.

From the [xds protocol](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#ack), the initial request is always empty.

lita commented on internal/app/upstream/client.go:
```diff
@@ -55,11 +74,177 @@ type Response struct {
+			request.VersionInfo = sig.version
+			err := doWithTimeout(func() error {
+				return stream.SendMsg(request)
+			}, callOptions.Timeout)
```
shouldn't the context provide the Deadline here?

jyotimahapatra replied:
the ctx does not have a deadline at this point. `doWithTimeout ` uses a child context with deadline.

lita replied:
I don't see where it uses the child context? the context doesn't seem to be passed into the doWithTimeout method.

jyotimahapatra replied:
At your time of comment i was using a timer to achieve timeout. I. have since changed the [code](https://github.com/envoyproxy/xds-relay/pull/46/files#diff-c30e80dfa8b61ea3ccdc1ae6d3716890R226) to use ctx and deadline. The github view here is probably stale.

jessicayuen commented on internal/app/upstream/client.go:
```diff
@@ -55,11 +74,177 @@ type Response struct {
+	}
+
+	signal := make(chan *version, 1)
+	signal <- &version{}
```
Can you explain what we're populating in this channel initially?

jyotimahapatra replied:
the initial seed is empty version and nonce. I should explicitly initialize them to make the intent clear and add comments.

jessicayuen commented on internal/app/upstream/client.go:
```diff
@@ -55,11 +74,177 @@ type Response struct {
+
+	go func() {
+		wg.Wait()
+		closeChannel(signal, response)
```
nit: s/closeChannel/closeChannels

jessicayuen commented on internal/app/upstream/client.go:
```diff
@@ -55,11 +74,177 @@ type Response struct {
+	callOptions CallOptions) {
+	for {
+		select {
+		case sig, ok := <-signal:
```
For my understanding, we'll only ever have one signal in the channel right? When the stream is first opened? (because orchestrator only calls `OpenStream` once per aggregated key). Asking because I want to understand the reasoning for the for loop here a bit more.

lita commented on internal/app/upstream/client_test.go:
```diff
@@ -0,0 +1,393 @@
+	)
+}
+
+func createMockLdsClient(
```
Rather than doing this, I think creating an internal/ or test/utils/fake package with mock xDS clients would be useful unless this is the only module requiring mocked out xDS clients.

jyotimahapatra replied:
I didn't take the approach since xdsclient mock is used only in one place.

jessicayuen replied:
I think orchestrator could use these rather than reimplementing. These will be valuable to be shareable imo.

jyotimahapatra replied:
I think the mock needed by orchestrator is bit different. It wouldn't need the complex setup provided by this particular mock in discussion. The mock needed by orchestrator would only need a simple `OpenStream` mock for easy testing. WDYT?

jyotimahapatra replied:
I have moved the mock into a reusable package. LMK how does it look.

jessicayuen commented on internal/app/upstream/client.go:
```diff
@@ -55,11 +74,177 @@ type Response struct {
+				}
+			}
+		case <-ctx.Done():
+			_ = stream.CloseSend()
```
Is `_` an error, should we handle it?

jyotimahapatra replied:
From [here](https://github.com/grpc/grpc-go/blob/master/stream.go#L786) looks the err is always nil and non actionable.

jessicayuen replied:
Interesting. Maybe just `stream.CloseSend()` then? (rather than assigning)

jyotimahapatra replied:
that is an artifact of linter:
```
internal/app/upstream/client.go:208:20: Error return value of `stream.CloseSend` is not checked (errcheck)
			stream.CloseSend()
```

jessicayuen commented on internal/app/upstream/client.go:
```diff
@@ -55,11 +74,177 @@ type Response struct {
+			default:
+				response <- &Response{Err: err}
+			}
+			return
```
How come we break out in the error scenario rather than trying to receive future messages?

jyotimahapatra replied:
Both [SendMsg](https://github.com/grpc/grpc-go/blob/master/stream.go#L715) and [RecvMsg](https://github.com/grpc/grpc-go/blob/master/stream.go#L734) perform the retries based on the grpc retry params. Since they are already performing the retries, i think we wouldn't benefit from retrying on top of them.

As i have put in `TODO` figuring out grpc retry settings will be a separate PR. When an error happens at this point, the best course would be to open a new stream.

jyotimahapatra commented on internal/app/server/server.go:
```diff
@@ -33,7 +33,7 @@ func Run() {
 
 	// Initialize upstream client.
-	upstreamClient, err := upstream.NewClient(ctx, upstreamClientURL)
+	upstreamClient, err := upstream.NewClient(ctx, upstreamClientURL, upstream.CallOptions{})
```
i wouldn't heap more on the already open PR. I will separately add the configurability.

jyotimahapatra commented on internal/app/upstream/client.go:
```diff
@@ -55,11 +74,177 @@ type Response struct {
-// TODO: pass retry/timeout configurations
-func NewClient(ctx context.Context, url string) (Client, error) {
-	return &client{}, nil
+// TODO: pass retry configurations
```
yeah. this was supposed to be deleted.

jyotimahapatra commented on internal/app/upstream/client.go:
```diff
@@ -55,11 +74,177 @@ type Response struct {
-	return &client{}, nil
+// TODO: pass retry configurations
+func NewClient(ctx context.Context, url string, callOptions CallOptions) (Client, error) {
+	conn, err := grpc.Dial(url, grpc.WithInsecure())
```
That is true. However, i'm not sure if that is everything that needs considering. I wouldn't consider the configurability of grpc params in this PR.

jessicayuen replied:
Sounds good, can you open a issue too?

jyotimahapatra commented on internal/app/upstream/client.go:
```diff
@@ -55,11 +74,177 @@ type Response struct {
+	case endpointTypeURL:
+		stream, err = m.edsClient.StreamEndpoints(ctx)
+	default:
+		return nil, fmt.Errorf("Unsupported request type")
```
good point. I will change this to a different error type. Depending on strings probably won;t be strong.

jyotimahapatra commented:
@jessicayuen @lita @eapolinario @LisaLudique PTAL. I have incorporated the feedback from the discussion and changed the interface. 
I would like to merge this and then work with the orchestrator and iterate on the interface as needed.

jyotimahapatra commented on internal/app/upstream/client_test.go:
```diff
@@ -0,0 +1,225 @@
+	})
+	assert.NotNil(t, respCh)
+	assert.Nil(t, err)
+	close(done)
```
one oddity of the ownership of done channel is that the channel is created in the xdsclient, but expects to be cleaned up in the caller.  We should solve it by changing the ownership of the channel towards the caller, but we can do it in a later iteration once we see value in it.

jessicayuen commented on internal/app/upstream/client.go:
```diff
@@ -55,11 +73,184 @@ type Response struct {
+				cancelFunc()
+				wg.Done()
+			default:
+				close(response)
```
Similar here, if this is irrecoverable, should we not just issue wg.Done() too? Do we need the select?

jessicayuen commented on internal/app/upstream/client.go:
```diff
@@ -55,11 +73,184 @@ type Response struct {
+	response := make(chan *v2.DiscoveryResponse, 1)
+
+	var wg sync.WaitGroup
+	wg.Add(2)
```
What's the reason that we wait on both recv and send to fail (in the case of an error) or succeed (in the case of a downstream signal to terminate), rather than just either/or? Can this just be `wg.Add(1)`?

jyotimahapatra replied:
The waitgroup guards the channels which are used in both send and recv goroutines. We need both goroutines to exit before we close the channels. We could block on both channels separately, but both in error and success cases we need both channels to be closed and hence they are together in one group.

jessicayuen commented on internal/app/upstream/client.go:
```diff
@@ -25,28 +45,26 @@ type Client interface {
+	OpenStream(*v2.DiscoveryRequest) (<-chan *v2.DiscoveryResponse, chan bool, error)
 }
 
 type client struct {
```
Would also add the logger to this struct and log on the error received upstream since we're no longer propagating downstream.

jyotimahapatra commented on internal/app/server/server.go:
```diff
@@ -38,7 +39,7 @@ func Run(bootstrapConfig *bootstrapv1.Bootstrap,
 	upstreamPort := strconv.FormatUint(uint64(bootstrapConfig.OriginServer.Address.PortValue), 10)
 	upstreamAddress := net.JoinHostPort(bootstrapConfig.OriginServer.Address.Address, upstreamPort)
-	upstreamClient, err := upstream.NewClient(ctx, upstreamAddress)
+	upstreamClient, err := upstream.NewClient(ctx, upstreamAddress, upstream.CallOptions{Timeout: time.Minute})
```
i think i wouldn't address this now since it is unrelated to the core working of the client. I will create an issue and link here.

jyotimahapatra commented on internal/app/upstream/client.go:
```diff
@@ -55,11 +73,184 @@ type Response struct {
+	signal <- &version{nonce: "", version: ""}
+
+	done := make(chan bool, 1)
+	response := make(chan *v2.DiscoveryResponse, 1)
```
This is intentional. We are creating a buffered channel since we dont want the goroutines to block on each other. Also, the send and recv calls are lock step and should not happen independently, hence 1. 
Some refs: [ref](https://medium.com/a-journey-with-go/go-buffered-and-unbuffered-channels-29a107c00268), [ref](https://www.ardanlabs.com/blog/2014/02/the-nature-of-channels-in-go.html)

jyotimahapatra commented on internal/app/upstream/client.go:
```diff
@@ -55,11 +73,184 @@ type Response struct {
+			if !ok {
+				return
+			}
+			request.ResponseNonce = sig.nonce
+			request.VersionInfo = sig.version
```
makes sense, will remove the pointer. Although fwiw, the pointer is not changing, so the orchestrator map will be unaffected.

jyotimahapatra commented on internal/app/upstream/client.go:
```diff
@@ -55,11 +73,184 @@ type Response struct {
+			if err != nil {
+				select {
+				case <-done:
+					cancelFunc()
```
There is a nuance here which is coming up because of using closed channel for error signaling: Closing channel on error vs closing channel cleanly.
The `done` channel indicates there was a signal from orchestrator to cleanup. It's not an error condition anymore. The orchestrator has to behave in 2 ways, if its signaling a done, it should not wait on a closed Response channel, because it will be anyway cleanly closed. If there was no signal via done to cleanup and the Response channel was still closed, it indicates an error. 
I was of the opinion that there should be a different signaling mechanism just to indicate error, lmk what you feel about it.

jyotimahapatra replied:
I might have misunderstood the question. I changed the implementation some more. LMK if that is aligned with what you are thinking.

jyotimahapatra commented:
@jessicayuen @lita @eapolinario @LisaLudique Please take another pass. I think its looking much better now.

lita commented on internal/app/upstream/client.go:
```diff
@@ -55,11 +74,179 @@ type Response struct {
+	}()
+	select {
+	case <-deadlineCtx.Done():
+		cancel()
```
This cancel context is not passed anywhere..not sure what this does.

jyotimahapatra replied:
i was flummoxed by this error message. Since the cancel operation is idempotent, and `deadlineCtx.Done()` basically means the deadline context is already cancelled, i used the cancel to be a noop and also to get rid of the error.
<img width="587" alt="Screen Shot 2020-04-17 at 9 00 31 AM" src="https://user-images.githubusercontent.com/49211422/79589759-c9efce80-808a-11ea-9e7c-f98be689d146.png">

lita replied:
I see. If you are calling cancel, I would just `defer cancel()` so you don't have to call it everywhere. `cancel()` is needed to release the resources it is associated with so it cleans itself up.

Also, does it make sense to use context.WithTimeout so it sends a timeout error? If not, then I would rename the private method to be `doWithDeadline`.

jyotimahapatra replied:
`context.WithTimeout` makes more sense. Will change `WithDeadline` to `WithTimeout`

lita commented on internal/app/upstream/client_test.go:
```diff
@@ -0,0 +1,221 @@
+
+type CallOptions = upstream.CallOptions
+
+func TestOpenStreamShouldReturnErrorForInvalidTypeUrl(t *testing.T) {
```
I don't see a test here for when the request times out and gets a DeadlineExceeded error

jyotimahapatra replied:
This one is the test: https://github.com/envoyproxy/xds-relay/blob/d3a64bf5d0ec08dd370804a29b820b1a858765e2/internal/app/upstream/client_test.go#L197-L219

We had decided that we don't need to bubble up the error to orchestrator and instead just close the channel.

lita replied:
But how do you assert that the error was due to a timeout vs if the code changes? Normally, I would test this via a stat and assert that the timeout stat was incremented. 

Can you add a TODO to add that?

jyotimahapatra replied:
Yeah i understand and i previously I meant to expose the error to the caller instead of just closing the channel. Since we have decided to close the channel to report all errors, we have to resort to alternate means to see which error really happens. I will add a TODO to check on a stat once stat is available.

jyotimahapatra replied:
added todo.

lita commented on test/mocks/upstream/client.go:
```diff
@@ -0,0 +1,202 @@
+package upstream
```
Rather than putting this in a giant mock repo, I think it might be better to put this in the same repo as upstream package, like so:

`internal/app/upstream/mock.go`
or
`internal/app/upstream/mock/client.go`

Modules using this mock will still need to import `"github.com/envoyproxy/xds-relay/internal/app/upstream"` anyway to use it. Might as well put them together.

jyotimahapatra replied:
Sure can do that, i think i moved this to mock package out so that other packages can use based on a previous comment. I like how explicitly other packages can import mock [here](https://github.com/envoyproxy/xds-relay/blob/0ac6a265387b46e81d5f10805ef06e1f8398b82d/internal/app/orchestrator/orchestrator_test.go#L11) [here](https://github.com/envoyproxy/xds-relay/blob/0ac6a265387b46e81d5f10805ef06e1f8398b82d/internal/app/upstream/client_test.go#L13).
LMK if you feel strongly about moving the file into the paths you suggested.

lita replied:
You can achieve the same result by putting the mocks in the `/mock` folder.

Putting the mock with the struct it is mocking makes it so that the dependencies are clear. If you are not careful, you can also end up with circular dependencies by putting them in one module.

jyotimahapatra replied:
sounds good. I've moved the mocks into `internal/app/upstream/mock`.

lita commented on internal/app/upstream/client.go:
```diff
@@ -55,11 +74,179 @@ type Response struct {
+				logger.Error(ctx, "Error in RecvMsg %s", err.Error())
+			}
+			cancelFunc()
+			closeChannels(signal, response)
```
I think it will be cleaner if you break and closeChannels at the end of this method.

lita replied:
I would also call this `shutdown` or `close`, as we might need to add extra steps.

jyotimahapatra replied:
makes sense. will do

lita commented on internal/app/upstream/client.go:
```diff
@@ -55,11 +74,179 @@ type Response struct {
+		}
+		select {
+		case <-ctx.Done():
+			closeChannels(signal, response)
```
same. break out of the for-loop so that the cleanup method happens in one place.

lita commented on internal/app/upstream/client.go:
```diff
@@ -55,11 +74,179 @@ type Response struct {
+			if err != nil {
+				select {
+				case <-ctx.Done():
+					// Context was cancelled, hence this is not an erroneous scenario.
```
Just so I understand, this only can happen during a timeout?

jyotimahapatra replied:
A timeout from `doWithTimeout` will will cause an error [here](https://github.com/envoyproxy/xds-relay/blob/0ac6a265387b46e81d5f10805ef06e1f8398b82d/internal/app/upstream/client.go#L242).
In the test cases i figured there's racecondition and shutdown could be called when timeout happens. In that case since shutdown causes the parent context cancellation, its prudent to check on it and return. Since the shutdown is already called, there is no point in reporting an error.

jyotimahapatra replied:
In other words throughout the implementation `ctx.Done()` happens only when `shutdown()` has been called. When goroutines get unblocked and find shutdown has been called, they should exit instead of continuing. I will put the intent in a comment.

jyotimahapatra replied:
added some more comments [here](https://github.com/envoyproxy/xds-relay/blob/f5a366b4816689977fc082da1e07a2f71377cf99/internal/app/upstream/client.go#L48-L50).

jyotimahapatra replied:
I think this is the remaining one. @lita lmk what you feel about the explanation.

lita replied:
I would add a comment around that the parent context was canceled, probably due to shutdown.

It would also be good to abstract this in a method as these nested selects are hard to read.

```
// Handles the scenario when sendMsg has an error
func handleSendErr(ctx context.Context, err error.Error, cancelFunc func) {
	defer cancelFunc()

	select {
	// If the context is done, then we are probably shutting down. Don't log the error and continue with shutdown.
	case <- ctx.Done():
		return 
	}

	logger.Error(ctx, "Error in SendMsg: %s", err.Error())

}
```

lita replied:
Looks like you have the same logic in recv as well which can leverage this code:

https://github.com/envoyproxy/xds-relay/pull/46/files#diff-c30e80dfa8b61ea3ccdc1ae6d3716890R198-R204

jyotimahapatra replied:
pulled the common piece into a function.

lita commented on internal/app/upstream/client_test.go:
```diff
@@ -0,0 +1,223 @@
+
+type CallOptions = upstream.CallOptions
+
+func TestOpenStreamShouldReturnErrorForInvalidTypeUrl(t *testing.T) {
```
You don't have to do this now, but in the future, it might be good to parameterized these tests as you always have to create a MockClient. The only parameters that are changing are CallOptions and the function being passed int.

jyotimahapatra replied:
done. There were 2 buckets of unique types of creation across 6 tests and I pulled them into separate functions.

lita commented:
Last set of comments, otherwise looks LGTM. will approve once that select statement is cleaned up.

lita commented on internal/app/upstream/client.go:
```diff
@@ -55,11 +77,177 @@ type Response struct {
+	defer cancelFunc()
+	select {
+	case <-ctx.Done():
+		// Context was cancelled, hence this is not an erroneous scenario.
```
Shouldn't this be above the case? 🤔 Not sure about go formatting so I could be wrong!

jyotimahapatra replied:
hm...i will look at why lint is not catching it in the next PR.
